### PR TITLE
Remove old parser

### LINKED ISSUE #59: Retire the old parser
Replace `TinyLang.Field.Parser` by a combination of `TinyLang.Field.Raw.Parser` and `TinyLang.Field.Typed.TypeChecker`.

### PR DESCRIPTION
**Pull request by jakzale**
Fixes #59 

- keeps the old API in `TinyLang.Field.Typed.Parser`, and
- generalises `TinyLang.Field.Raw.Parser`.

### CODE REVIEW

File: test/Field/golden/forLoops.golden [hunk 1]
```diff
@@ -1 +1 @@
-let i'_6 = 1; let k_7 = 1 * 2; assert k_7 == (i'_6 * 2); let k_8 = 1 * 3; assert k_8 == (i'_6 * 3); let p_9 = 1; let p_10 = p_9 * 1; let p_11 = p_10 * 2; let i'_12 = 2; let k_13 = 2 * 2; assert k_13 == (i'_12 * 2); let k_14 = 2 * 3; assert k_14 == (i'_12 * 3); let p_15 = 2; let p_16 = p_15 * 1; let p_17 = p_16 * 2; p_17
\ No newline at end of file
+let i'_5 = 1; let k_6 = 1 * 2; assert k_6 == (i'_5 * 2); let k_7 = 1 * 3; assert k_7 == (i'_5 * 3); let p_8 = 1; let p_9 = p_8 * 1; let p_10 = p_9 * 2; let i'_11 = 2; let k_12 = 2 * 2; assert k_12 == (i'_11 * 2); let k_13 = 2 * 3; assert k_13 == (i'_11 * 3); let p_14 = 2; let p_15 = p_14 * 1; let p_16 = p_15 * 2; p_16
```
Comment (jakzale) [inline test/Field/golden/forLoops.golden]: @effectfully for some reason the unique/id generation went down by one.

File: field/TinyLang/Field/Raw/Parser.hs [hunk 2]
```diff
@@ -171,7 +170,7 @@ import           Data.Set ( fromList
                           , member
                           )
 
-type Parser = Parsec Void String
+type Parser' = ParsecT Void String
```
Comment (effectfully) [inline field/TinyLang/Field/Raw/Parser.hs]: Let's call it `ParserT` then.

File: field/TinyLang/Field/Typed/Parser.hs [hunk 3]
```diff
@@ -73,18 +73,16 @@ module TinyLang.Field.Typed.Parser
 
 import           TinyLang.Prelude               hiding (many, some, try, option)
```
Comment (effectfully) [inline field/TinyLang/Field/Typed/Parser.hs:16]: Remove the docs at the beginning of the file and just refer to the raw parser.

File: test/Field/golden/forLoops.golden (hunk 1 above)
Comment (effectfully) [reply test/Field/golden/forLoops.golden]: This is super weird. If I remove `renameExpr` from `parseScopedExpr` in the previous version of the parser, I get:

```
let i'_1 = 1; let k_3 = 1 * 2; assert k_3 == (i'_1 * 2); let k_3 = 1 * 3; assert k_3 == (i'_1 * 3); let p_4 = 1; let p_4 = p_4 * 1; let p_4 = p_4 * 2; let i'_1 = 2; let k_3 = 2 * 2; assert k_3 == (i'_1 * 2); let k_3 = 2 * 3; assert k_3 == (i'_1 * 3); let p_4 = 2; let p_4 = p_4 * 1; let p_4 = p_4 * 2; p_4
```

and if I remove it in the new version, I get

```
let i'_1 = 1; let k_3 = 1 * 2; assert k_3 == (i'_1 * 2); let k_3 = 1 * 3; assert k_3 == (i'_1 * 3); let p_4 = 1; let p_4 = p_4 * 1; let p_4 = p_4 * 2; let i'_1 = 2; let k_3 = 2 * 2; assert k_3 == (i'_1 * 2); let k_3 = 2 * 3; assert k_3 == (i'_1 * 3); let p_4 = 2; let p_4 = p_4 * 1; let p_4 = p_4 * 2; p_4
```

which look identical to me. How come the renamer renames two identical expressions differently? We didn't seem to change it, or at least I do not see any changes there. Weird.

Anyway, the result of the renaming looks correct.

Could you also change `TinyLang.Field.Typed.Core.supplyFromAtLeastFree` to

```haskell
supplyFromAtLeastFree :: MonadSupply m => Expr f a -> m ()
supplyFromAtLeastFree =
    supplyFromAtLeast . freeUniqueIntMap . unEnv . _scopedVarSignsFree . exprVarSigns
```

? We shouldn't be bothered about bound variables there.

... Aaahh! In `runTypeChecker` you do `runSupplyT`, so type checking leaves the `Supply` monad, so when we start doing renaming generation of `Unique`s starts with `0` again.

The Barendregt's convention is a terrible minefield. I find it more reliable to defer `runSupply`/`runSupplyT` as closer to the call site as possible. So I suggest to make

```haskell
runTypeChecker :: (MonadError TypeCheckError m, MonadSupply m) => TypeChecker a -> m a
```

It won't have much effect here, but it's generally a more reliable strategy to follow.

**Review by effectfully — APPROVED** [review]

File: test/Field/golden/forLoops.golden (hunk 1 above)
Comment (jakzale) [reply test/Field/golden/forLoops.golden]: Fixing `supplyFromAtLeastFree` restarts `Uniques` from 0.

I will update the `runTypeChecker` per your suggestion.

Comment (jakzale) [reply test/Field/golden/forLoops.golden]: @effectfully, does it also mean that `TypeChecker.typeCheck` and `TypeChecker.checkType` should also be exposed with `MonadError TypeCheckError m, MonadSupply m)` constraint?

Comment (effectfully) [reply test/Field/golden/forLoops.golden]: Yes. Thanks, didn't notice that.

`runSupply`/`runSupplyT` should be deferred to the call site where possible. This is the general strategy that we follow on the Plutus side and it allowed us to reduce overall pain of name handling.

Unless it's way too annoying, e.g. wrapping every single term in `Supply` is something that we originally started with, but it proved to be irritating without actually giving much safety benefit.

Comment (effectfully) [reply test/Field/golden/forLoops.golden]: > Fixing `supplyFromAtLeastFree` restarts Uniques from `0`.

Yeah, it was always supposed to do that, I'm not sure how I ended up also counting bound variables there. Maybe an experiment, maybe some general paranoia.

**Pull request merged**